### ISSUE TITLE
Use Moo (with MooX::Singleton and Type::Tiny) as alternative for Moose

### ISSUE THREAD

**Issue by nicolasfranck** [localh0rst/OpenSearch-Perl#7]
Would it be feasible to use `Moo` as package constructor, in combination with `Type::Tiny` for parameter validation?

I see a lot of object construction calls for every search operation (the `execute` statements).
Moose also a slower startup than Moo (due to a higher complexity).

For the parameter packages (`OpenSearch::Parameters::*`) you would then need a different of course. Not directly an idea of where to store extra information in Moo to be used by the helper methods.

`Type::Tiny` has an optional package `Type::Tiny::XS` for speedup.

When not feasible, I know of a few old tricks that I used in the past like adding 

```
__PACKAGE_->meta->make_immutable();
```

at the end of every Moose package. See also https://metacpan.org/dist/Moose/view/lib/Moose/Cookbook/Basics/Immutable.pod

I am willing to give it a try ;-), but we need to be on the same page here.

**Comment by localh0rst** [comment]
Im still not sure if "local" parameter validation is necessary at all (except a check for allowed, required and optional parameters). Its just a tradeoff between local validation vs "remote" validation. My thought was: "

Why should i send a request if i know its wrong"

However, just like ''track_total_hits" there are many parameters that allow "multiple types" (i.e. '_source'). Checking for each of these might not be a good idea and very time-consuming (i was trying to do this via "Filters". i.e OpenSearch::Filter::Source, but i abandoned this approach). 

Even though i would be okay to use Moo with `Type::Tiny` as a replacement this might get a little tricky. All Parameters
currently require MooseX::MetaDescription::Meta::Attribute where some magic happens. The description key is currently used by `OpenSearch::Helper::_generate_params()` to determine required body/url parameters and if they are required or not (and some encode_functions). 

Using '__PACKAGE__->meta->make_immutable` would be the easier approach for now.

If you have any idea how to make parameter-validation better (there sure is a way) im very much willing to go that route ;)

**Comment by nicolasfranck** [comment]
* About the local parameter validation: there is some truth in using local validation. Other languages have the luxury of having static types that are checked on compilation (so you cannot supply an invalid type). But then again, checking too much has the disadvantage that the library must follow every evolution in the api (although that may be desirable; not worse than undocumented return values). So I'm not truly convinced to leave this path altogether. You're right. Without local validation it would become more complex to encode the values.
* About `Type::Tiny`: the only problem is indeed in that Parameter namespace where you use the meta property of Moose to document how and where to encode the attribute in the request. I was thinking of adding a required  method `to_request` (and move the whole package into its client package) or `encode` method in every Parameter package. It is more work indeed (`if ($self->index) { ... }`), but more readable?

**Comment by localh0rst** [comment]
This might be a better approach. Could you provide some more info (maybe a small example) on how 
you would accomplish this?

Ill try to find some time and also think about this problem.

**Comment by nicolasfranck** [comment]
I'll provide an example tomorrow. I can also help.

**Comment by localh0rst** [comment]
Thank you.

As a (small) start i added __PACKAGE__->meta->make_immutable (in the development branch) to all Moose classes.
As stated in the documentation, it comes with a small compile-time cost (470ms vs 520ms). I did not check object-
creation speed.

**Comment by nicolasfranck** [comment]
I also had a situation where the singleton of `OpenSearch::Base` could not be created. MooseX::Singleton reported that the instance was already created. I tried to replace it by (example):

```
package OpenSearch::Base;

my $instance;

sub set_instance ($class, $self) {
   $instance = $self;
}
sub instance ($class) {
  $instance;
}

1;
```

```
package OpenSearch;

...

sub BUILD {
  ..
  OpenSearch::Base->set_instance($self->base);
}
```

and then it started working again. Not sure how to replicate the problem

**Comment by nicolasfranck** [comment]
About the place where one could store encoding information per arguments: one could make a single hash that documents the whole api.

cf. https://github.com/nics/WWW-ORCID/blob/main/lib/WWW/ORCID/API/v2_0.pm#L14 for ORCID v2.0

The key could be the path (e.g. `/_search`), followed by a hash as value that lists the available parameters, the encoding function, and the place to encode:

```
my $opts = +{
    "/_search" => {
     index => {
       type => 'path',
     },
     q => {
       type => 'query',
     },
      refresh => {
        type => 'query',
        encode_func => \&encode_bool,
      }
   }
};
```

or like this: https://github.com/elastic/elasticsearch-perl/blob/f427713b8f398fe6738dc5e5d547673786f92dd1/lib/Search/Elasticsearch/Client/7_0/Role/API.pm#L2358

**Comment by localh0rst** [comment]
Tbh. im not a fan of the Elasticsearch way. Are we still on the same page and we want to replace Moose with Moo and leave the current `OpenSearch::Parameters::*` as it is and replace the `MooseX::MetaDescription::Meta::Attribute` stuff?

Maybe one could add a additional attribute to the Parameter Role. i.e. `__api_definition`:

```
package OpenSearch::Parameters::Search::Search;
use Moo::Role;

has 'index' => (is => 'rw', isa => 'Str')
...
has  '__api_definition' => (
  is => 'ro',
  default => sub {return {
    index => {encode_func => 'as_is, type => 'path', required => 0},
    q => {encode_func => 'as_is', type => 'url'}
  }}
)
```

What do you think about that? I really dont know whats the best way is.

**Comment by nicolasfranck** [comment]
The parameter package is fine. Maybe cleaner to read.

About the api specification, why not use one of the following:

* a package variable:

```
our %spec = (
  index => {encode_func => 'as_is, type => 'path', required => 0},
  q => {encode_func => 'as_is', type => 'url'
)
```

* a method that returns the specification as a hash reference:

```
sub api_spec {
  state $s = +{
    index => {encode_func => 'as_is, type => 'path', required => 0},
    q => {encode_func => 'as_is', type => 'url'
  };
}
```

That way you distinguish parameters for opensearch from internal attributes

btw it is still your module, I'm glad someone is working on it ;-)

**Comment by localh0rst** [comment]
I think its best to have a hash containing all Information and then generating all attributes based on that hash. The following test seems to work:

```
package Test::Role::Attr;
use Moo::Role;
use Type::Tiny::XS;
use Types::Standard qw( Str Int Enum ArrayRef Object );
use Data::Dumper;
my $attrs = {
  index => {
    is          => 'rw',
    isa         => Str,
    encode_func => 'as_is',
    type        => 'path',
    required    => 0
  },
  sort => {
    is          => 'rw',
    isa         => Str,
    encode_func => 'as_is',
    type        => 'body',
    required    => 0
  },
};
foreach my $attr ( keys %$attrs ) {
  has $attr => (
    is  => $attrs->{$attr}->{is},
    isa => $attrs->{$attr}->{isa},
  );
}
sub my_attrs {
  return $attrs;
}

package Test;
use Moo;
with 'Test::Role::Attr';

package main;
my $test  = Test->new( index => 'test', sort => 1 );
my $test2 = Test->new( index => 'test', sort => 1 );
my $spec = $test->my_attrs();
```

Even its only a little test, startup time seems to be halfed compared to using Moose

**Comment by nicolasfranck** [comment]
I'll do some tests. My ideas where kind of the same

**Comment by localh0rst** [comment]
Would you mind testing the convert-to-moo branch? At least `minil test` seems to run all tests okay

**Comment by nicolasfranck** [comment]
yes of course; I'll send a notice when ready

**Comment by localh0rst** [comment]
Fixed by #15 

Thanks for fixing your own issue ;)
I`ll test it some more and release it to CPAN.

**Comment by localh0rst** [comment]
closed
